rust/osm2lanes/src/transform/tags_to_lanes/lane.rs

```diff
@@ -0,0 +1,266 @@
+                    // TODO lanes.downgrade(&[forward, backward, bothways]);
+                    Self {
+                        lanes,
+                        forward,
+                        backward,
+                        bothways,
+                    }
+                },
+                (Some(l), Some(f), None) => Self {
+                    lanes: Infer::Direct(l),
+                    forward: Infer::Direct(f),
+                    backward: Infer::Calculated(l - f - bothway_lanes),
+                    bothways,
+                },
+                (Some(l), None, Some(b)) => Self {
+                    lanes: Infer::Direct(l),
+                    forward: Infer::Calculated(l - b - bothway_lanes),
+                    backward: Infer::Direct(b),
+                    bothways,
+                },
+                // Alleyways or narrow unmarked roads, usually:
+                (Some(1), None, None) => Self {
+                    lanes: Infer::Direct(1),
+                    forward: Infer::Default(0),
+                    backward: Infer::Default(0),
+                    bothways: Infer::Guessed(1),
+                },
+                (Some(l), None, None) => {
+                    if l % 2 == 0 && centre_turn_lane.present.some().unwrap_or(false) {
+                        // Only tagged with lanes and deprecated center_turn_lane tag.
+                        // Assume the center_turn_lane is in addition to evenly divided lanes.
+                        Self {
+                            lanes: Infer::Calculated(l + 1),
+                            forward: Infer::Guessed(l / 2),
+                            backward: Infer::Guessed(l / 2),
+                            bothways: Infer::Calculated(1),
+                        }
+                    } else {
+                        // Distribute normal lanes evenly.
+                        let remaining_lanes =
+                            l - bothway_lanes - forward_bus_lanes - backward_bus_lanes;
+                        if remaining_lanes % 2 != 0 {
+                            warnings.push(TagsToLanesMsg::ambiguous_str("Total lane count cannot be evenly divided between the forward and backward"));
+                        }
+                        let half = (remaining_lanes + 1) / 2; // usize division rounded up.
+                        Self {
+                            lanes: Infer::Direct(l),
+                            forward: Infer::Guessed(half + forward_bus_lanes),
+                            backward: Infer::Guessed(
+                                remaining_lanes - half - bothway_lanes + backward_bus_lanes,
+                            ),
+                            bothways,
+                        }
+                    }
+                },
+            }
+        }
+    }
+}
+
+const CENTRE_TURN_LANE: TagKey = TagKey::from("centre_turn_lane");
+pub struct CentreTurnLaneScheme {
+    pub present: Infer<bool>,
```

Comment:
what does "present" mean in this context?

Reply:
It refers to the presence of such a lane: "is a centre turn lane present on this road?".

Is something like `struct CentreTurnLaneScheme(Infer<bool>)` possible? I haven't used that language feature before.

Reply:
yes, that would be the correct construct.

Reply:
https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types